### ISSUE TITLE
[QST] Decoding item-id in NextItemPredictionTask

### ISSUE THREAD

**Issue by samhedin** [NVIDIA-Merlin/Transformers4Rec#612]
# ❓ Questions & Help

[02-end-to-end-session-based-with-Yoochoose-PyT](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/end-to-end-session-based/02-End-to-end-session-based-with-Yoochoose-PyT.ipynb) describes how to produce recommendations for items.

At the bottom, the encoded item id's are given as predictions:
```
- Top-5 predictions for session `11394056`: 6510 || 7055 || 4136 || 4204 || 4541

As you noticed, we first got prediction results (logits) from the trained model head, and then by using a handy util function visualize_response we extracted top-k encoded item-ids from logits.
```
If I understand correctly, these item-ids are encoded. How do I get the original item-id back?
The encoding is done with nvtabular Categorify, similar to here: https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/end-to-end-session-based/01-ETL-with-NVTabular.ipynb
In my case the original item-id is a string.

I would like tritonserver to return the decoded item-id's, since I will make the request from a server which I hope will not need to know anything about the internals of the model.

Thank you!

**Comment by rnyak** [comment]
@samhedin please refer to this [ticket](https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/511) where I explain what triton returns, incase you have set `start_index=1` in the Categorify op, and in case you dont. 

currently we dont have an automatic way to return to decoded item ids from Triton.  We are working on a wrapper and a utility function for that. but you can convert encoded triton ids to original ids following up the discussion in the [ticket](https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/359). what you need is the `unique_item_id.parquet` file in the categories folder.  you can create a simple mapping utility function to do post-processing decoding.

**Comment by samhedin** [comment]
Thanks for the quick response, using `unique_item_id.parquet` is easy enough!
One last question, in the linked ticket you say 
```
if you used start_index =1 in the Categorify() op then your item_id encoding starts from 2 not 1. Because we reserve 0 for padding, 1 for OOV +nulls, and then the other item categories starts from 2. so in that case, you will get 2 + n_unique_items from your prediction results.
```
I used `Categorify(start_index=1)`, meaning that I should get `n_unique_items + 2`.
My `unique_item_id.parquet` looks like this, it seems to lack 0 for padding.
```
0                                               <NA>         0
1                                            <item1>     4500
2                                             <item2>       1800
3                                             <item3>      1200
4                                             <item4>     1200
5                                              <item5>       700
6                                             <item6>       600
```
Which to me looks like `n_unique_items + 1`. If I produce some random recommendations on this tiny sample dataset, I get 
```
5
1
7
2
3
5
1
```
Which again seems like `n_unique_items + 1` (6 unique items, one OOV/null item).
Is the padding token not included?

And to be clear, to decode these, I would do something like (notice the -1 in df.iloc)
```
    df = pd.read_parquet("unique.item_id.parquet")
    for user, next_items in zip(users, top_preds):
        decoded = [df.iloc[e - 1] for e in next_items]
        print(f"User {user} recommendations: {decoded}")
```

Because the output from TIS contained a 7 index which is out of bounds.

**Comment by rnyak** [comment]
>I used Categorify(start_index=1), meaning that I should get n_unique_items + 2.
yes but not in the `unique_item_id.parquet` file..  you get `n_unique_items +1` in the unique parquet file.  you will get `+2` in your results coming from Triton server response.

so you only need to shift the index of the parquet file `+1` . if you used `start_index=1` , just  add +1 to the index column.

                                                  item_id    item_id_size
1                                               <NA>         0
2                                           <item1>     4500
3                                            <item2>       1800
4                                            <item3>      1200
5                                             <item4>     1200
6                                              <item5>       700
7                                             <item6>       600


So  you can do like that, get the scores from Triton. Note that first score belong to padding, second belong to OOV, third belong to encoded item id 2, fourth score belong to encoded item id 3, so on so forth.... 

so you can first sort the scores by preserving/storing their indices as well,  and then do the reverse mapping using the  logic above that I mentioned.

**Comment by samhedin** [comment]
Thank you so much!

**Comment by denisov-vlad** [comment]
Hi @samhedin 

I've written extra step in t4r ensemble which returns both items and scores:


**item_model.py** file for Triton. This file is copied by next function.

```python
import logging
import pathlib

import cudf
import triton_python_backend_utils as pb_utils


LOG = logging.getLogger("nvtabular")


class TritonPythonModel:
    """
    Triton model used for running NVT Workflows
    """

    def initialize(self, args):
        # Arg parsing
        repository_path = pathlib.Path(args["model_repository"])

        # Handle bug in Tritonserver 22.06
        # model_repository argument became path to model.py
        if str(repository_path).endswith(".py"):
            repository_path = repository_path.parent.parent

        items_path = (
            repository_path
            / str(args["model_version"])
            / "workflow/categories/unique.item.parquet"
        )
        model_device = args["model_instance_kind"]

        # Workflow instantiation
        self.items = cudf.read_parquet(items_path).item.to_numpy().reshape(1, -1)
        self.items = pb_utils.Tensor("items", self.items)

    def execute(self, requests):
        """Transforms the input batches by running through a NVTabular workflow.transform
        function.
        """
        responses = []
        for request in requests:
            probas = pb_utils.get_input_tensor_by_name(request, "proba")

            result = [probas, self.items]

            responses.append(pb_utils.InferenceResponse(result))

        return responses

```

It ads new model `t4r_item` and changes generated ensemble model. Maybe you need to fix some paths.

```python
import os
import pathlib
import shutil

import tritonclient.grpc
import tritonclient.grpc.model_config_pb2 as model_config
from google.protobuf import text_format


def prepare_item_model(model_path):
    workdir = os.environ.get("WORKDIR", "/opt/recs")

    item_model_path = f"{workdir}/train/models/t4r/src/item_model.py"

    t4r_path = f"{model_path}/t4r"
    t4r_nvt_path = f"{t4r_path}_nvt"
    t4r_item_path = f"{t4r_path}_item"
    categories_path = "1/workflow/categories/"
    t4r_categories_path = f"{t4r_item_path}/{categories_path}"
    item_parquet_path = f"{categories_path}/unique.item.parquet"
    ensemble_schema_path = f"{t4r_path}/config.pbtxt"
    ensemble_schema_path_bak = f"{ensemble_schema_path}.bak"

    schema = model_config.ModelConfig()

    save_ensemble_backup = True

    if os.path.exists(ensemble_schema_path_bak):
        ensemble_schema_read_path = ensemble_schema_path_bak
        save_ensemble_backup = False
    else:
        ensemble_schema_read_path = ensemble_schema_path

    with open(ensemble_schema_read_path, "r") as f:
        graph_str = f.read()
        ensemble_schema = text_format.Parse(graph_str, schema)

    output = ensemble_schema.output[0]

    # construct t4r_item config
    t4r_item = model_config.ModelConfig(name="t4r_item", backend="python")
    t4r_item.input.append(
        model_config.ModelInput(
            name="proba", data_type=output.data_type, dims=output.dims
        )
    )
    t4r_item.output.append(
        model_config.ModelOutput(
            name="items", data_type=model_config.TYPE_STRING, dims=output.dims
        )
    )
    t4r_item.output.append(
        model_config.ModelOutput(
            name="proba", data_type=output.data_type, dims=output.dims
        )
    )

    pathlib.Path(t4r_categories_path).mkdir(parents=True, exist_ok=True)
    shutil.copy2(f"{t4r_nvt_path}/{item_parquet_path}", t4r_categories_path)
    shutil.copy2(item_model_path, f"{t4r_item_path}/1/model.py")

    with open(f"{t4r_item_path}/config.pbtxt", "w", encoding="utf8") as o:
        text_format.PrintMessage(t4r_item, o)

    ensemble_schema.output.clear()
    ensemble_schema.output.MergeFrom(t4r_item.output)
    item_step = model_config.ModelEnsembling.Step(
        model_name="t4r_item", model_version=-1
    )
    item_step.input_map["proba"] = "output"
    item_step.output_map["proba"] = "proba"
    item_step.output_map["items"] = "items"

    ensemble_schema.ensemble_scheduling.step.append(item_step)

    if save_ensemble_backup:
        shutil.move(ensemble_schema_path, ensemble_schema_path_bak)

    with open(ensemble_schema_path, "w", encoding="utf8") as o:
        text_format.PrintMessage(ensemble_schema, o)
```
